### ISSUE TITLE
Compatibility layer for classic notebook extensions using the `IPython` / `Jupyter` globals

### ISSUE THREAD

**Issue by jtpio** [jupyterlab/retrolab#274]
### Problem

A lot of classic notebook extensions currently rely on `IPython` and `Jupyter` being exposed on `window`.

This makes porting extensions to JupyterLab / RetroLab a bit difficult since this is not supported anymore.

### Proposed Solution

Maybe RetroLab (and only RetroLab) should provide a compat layer and expose `window.IPython` and `window.Jupyter`. So things like `Jupyter.notebook.save()` executed in a JS cell still works in RetroLab by default.

An alternative for now is to use the `jupyterapp` main application on `window` with `--expose-app-in-browser`:

https://github.com/jupyterlab/retrolab/blob/6ce134fa17cc87d79fca822cbb761e4c7237b879/retrolab/app.py#L191-L195

Which can be used to execute arbitrary commands:

https://user-images.githubusercontent.com/591645/140728366-0883cf50-dba6-4542-b686-940116fd12ec.mp4

This can be implemented as a new extension in the retrolab repo. The extension would expose the classic notebook API, but call JupyterLab APIs internally.

### Additional context

This would help with use cases like https://github.com/ucbds-infra/otter-grader/issues/448

cc @yuvipanda @davidwagner @fperez @tonyfast

**Comment by fperez** [comment]
In general, I'm very positive to the idea of using retro as a place to experiment with what level of compatibility shims and new APIs would work well. It's a smaller, safer playground, and we can always promote into new core JupyterLab APIs anything deemed generic and robust enough.

This seems like an excellent idea, thanks @jtpio.

**Comment by psychemedia** [comment]
I don't really understand how JupyterLab / RetroLab extensions work so I'm not sure what this does or does not make possible, what any consequences are / what alternative approaches would be required to achieve what functionality provided by such a shim using current Jupyterlab/Retrolab API?

If the proposal was to work towards a way of putting a harness around the majority of current classic extensions, or providing a clear set of rules for how you could start to migrate different patterns from classic nb extensions to JupyterLab, that would give me a path I could start trying to walk my extensions down.

I also wonder what this does to the JupyterLab extension ecosystem? Would folk have to add tags to show whether their extension works only in JupyterLab, only in RetroLab, or in both?

And I wonder if these API *can* be exposed, why were they not exposed from the start? Or if they were, why were they withdrawn? What architectural or security principles are JupyterLab puritans being asked to give up by exposing these APIs?! ;-)

**Comment by jtpio** [comment]
> If the proposal was to work towards a way of putting a harness around the majority of current classic extensions, or providing a clear set of rules for how you could start to migrate different patterns from classic nb extensions to JupyterLab, that would give me a path I could start trying to walk my extensions down.

Yes this is mostly about buying us a bit more time, since we probably can't expect all extensions to migrate to JupyterLab for notebook 7. Also this issue is more about targeting simple use cases, such as `otter-grader` executing code like `display(Javascript("Jupyter.notebook.save_checkpoint();"))`, rather than complex extensions like nbgrader or RISE (for those it would be better to do a proper port). We don't need to cover the whole `Jupyter.notebook` API if it's not necessary.

> I also wonder what this does to the JupyterLab extension ecosystem? Would folk have to add tags to show whether their extension works only in JupyterLab, only in RetroLab, or in both?

We should still actively encourage developers to make JupyterLab extensions as this is the way forward. As discussed in other places, the tooling around authoring JupyterLab extensions can still be improved so it's more welcoming to non-developers. Folks should not start making new extensions relying on `IPython` / `Jupyter`. Actually if we add them to retro we could mark them as deprecated from the start. Because they will just be available to help with the transition.

> And I wonder if these API _can_ be exposed, why were they not exposed from the start? Or if they were, why were they withdrawn? What architectural or security principles are JupyterLab puritans being asked to give up by exposing these APIs?! ;-)

For several reasons I would say. Having `Jupyter.notebook` exposed on `window` is quite of a strong assumption specific to classic. In JupyterLab there might not even be a notebook displayed on the page.

**Comment by jtpio** [comment]
> experiment with what level of compatibility shims and new APIs

We should probably scope this to provide some kind of compatibility with old extensions only. New API development should be done in the main JupyterLab repo, so we don't fragment the lab ecosystem with new extensions that only work in RetroLab.

**Comment by yuvipanda** [comment]
Right now, there's no 'simple' way for python code to perform actions on the currently open notebook - something that a number of projects want to do. I think we can and should expose a safe subset of actions to this, and a global object accessible from `display` seem a good way to go. I do think it should be thought of as separate from a compatibility shim though - I don't want to have to install 200 extensions that all just simply let a python (or R or...) piece of code ask to save the notebook.

So for the otter case, instead of compatibility shims, we could instead expose something else that can be called via `display()` - just something that can perform a list of actions we have decided are safe for python code to do - save and checkpoint comes to mind, as is getting the *name* of the notebook. We can make these function calls that can somehow perhaps work with 'current notebook' even in lab - since this code goes into python libraries, they should ideally work across frontends. This can totally be just an extension to be installed, although I think for max compatibiltiy these should ship by default.

If needed, we can also search GitHub for code that calls `Jupyter` / `IPython` stuff via `display()` and provide backwards compat shims for that, but that can be a separate activity.

**Comment by jtpio** [comment]
> We can make these function calls that can somehow perhaps work with 'current notebook' even in lab - since this code goes into python libraries, they should ideally work across frontends. This can totally be just an extension to be installed, although I think for max compatibiltiy these should ship by default.

It can definitely be implemented as an extension. In fact some folks have already implemented some kind of proxy to do that in lab (https://github.com/jupyterlab/jupyterlab/issues/5660).

> This can totally be just an extension to be installed, although I think for max compatibiltiy these should ship by default.

By default in JupyterLab? Not sure. For example these wouldn't really apply to the code consoles.

**Comment by bollwyvl** [comment]
I'm still generally `-1` on re-hoisting a compatibility layer by default. Requirejs, jquery, jquery ui is going to be a slippery slope, and would firmly want to not see it on lab core.

However, unifying commands as a display type, and increasing the robustness of command definition e.g. https://github.com/jupyterlab/lumino/issues/58 sounds lovely.

A `application/jupyter-command+json` (sic) display type would be very handy, but i think would need some significant (user specified) guardrails. Basically, if i say, my kernel _can save this notebook_ shouldn't mean it can _delete all the files_. So the first time it asks, you'd get an in-line display (no modals, thx):

```
Kernel Python 3 would like to:

> Save Notebook Foo.ipynb

[ALLOW] [DENY] [REMEMBER...]
```

Where the "remember" would let you update your settings. In a managed hub setting, this could be configured with `overrides.json`.

**Comment by blink1073** [comment]
Colab offers a limited API for interacting between JS and Python: https://colab.research.google.com/notebooks/snippets/advanced_outputs.ipynb.  It seems like an appropriate scope for what could be offered in "Notebook 7".
